token.go

```diff
@@ -0,0 +1,280 @@
+// When both are present, SigningCertificateV2 is used because it uses hash
+// algorithm beyond SHA1. If both are missing, an error will be returned.
+//
+// References: RFC 3161 2.4.1 & 2.4.2; RFC 5816; RFC 5035 4; RFC 2634 5.4
+func (t *SignedToken) GetSigningCertificate(signerInfo *cms.SignerInfo) (*x509.Certificate, error) {
+	var signingCertificate signingCertificate
+	var expectSigningCertificateV1 bool
+	var signingCertificateV2 signingCertificateV2
+	if err := signerInfo.SignedAttributes.Get(oid.SigningCertificateV2, &signingCertificateV2); err != nil {
+		if !errors.Is(err, cms.ErrAttributeNotFound) {
+			return nil, fmt.Errorf("failed to get SigningCertificateV2 from signed attributes: %w", err)
+		}
+		// signingCertificateV2 is missing, try signingCertificate v1 instead
+		expectSigningCertificateV1 = true
+		if err := signerInfo.SignedAttributes.Get(oid.SigningCertificate, &signingCertificate); err != nil {
+			if !errors.Is(err, cms.ErrAttributeNotFound) {
+				return nil, fmt.Errorf("failed to get SigningCertificate from signed attributes: %w", err)
+			}
+			// signingCertificate v1 is missing as well
+			return nil, errors.New("invalid timestamp token: both signingCertificate and signingCertificateV2 fields are missing")
+		}
+	}
+	// get candidate signing certificate
+	var candidateSigningCert *x509.Certificate
+	var issuerSerial issuerAndSerial
+	if expectSigningCertificateV1 {
+		if len(signingCertificate.Certificates) == 0 {
+			return nil, errors.New("signingCertificate does not contain any certificate")
+		}
+		issuerSerial = signingCertificate.Certificates[0].IssuerSerial
+	} else {
+		if len(signingCertificateV2.Certificates) == 0 {
+			return nil, errors.New("signingCertificateV2 does not contain any certificate")
+		}
+		issuerSerial = signingCertificateV2.Certificates[0].IssuerSerial
+	}
+	signed := (*cms.ParsedSignedData)(t)
+	if issuerSerial.SerialNumber != nil {
+		// use IssuerSerial from signed attribute
+		if issuerSerial.IssuerName.Name.Bytes == nil {
+			return nil, errors.New("issuer name is missing in IssuerSerial")
+		}
+		var issuer asn1.RawValue
+		if _, err := asn1.Unmarshal(issuerSerial.IssuerName.Name.Bytes, &issuer); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal issuer name: %w", err)
+		}
+		ref := cms.IssuerAndSerialNumber{
+			Issuer:       issuer,
+			SerialNumber: issuerSerial.SerialNumber,
+		}
+		candidateSigningCert = signed.GetCertificate(ref)
+	} else {
+		// use sid (unsigned) as IssuerSerial
+		candidateSigningCert = signed.GetCertificate(signerInfo.SignerIdentifier)
+	}
+	if candidateSigningCert == nil {
+		return nil, CertificateNotFoundError(errors.New("signing certificate not found in the timestamp token"))
+	}
+	// validate hash of candidate signing certificate
+	var hashFunc crypto.Hash
+	var expectedCertHash []byte
+	if expectSigningCertificateV1 {
+		// Reference: https://datatracker.ietf.org/doc/html/rfc2634#section-5.4.1
+		hashFunc = crypto.SHA1
```

Comment:
I'm curious how many TSAs are still using SHA-1.

Reply:
From my testing, there is at least 1 free TSA that's still using SHA-1. This is not the signature algorithm of the certificate though, it's just the `certHash` field. 
[RFC 5816](https://datatracker.ietf.org/doc/html/rfc5816) says we should support both cases.